client/state/site-settings/exporter/actions.js

```diff
@@ -26,41 +43,168 @@ export function setPostType( postType ) {
- * Sends a request to the server to start an export.
+ * Sets one of the advanced export settings values in the UI
  *
+ * @param  {string} section   The name of the section containing the setting - 'posts', 'pages', or 'feedback'
+ * @param  {string} setting   The name of the setting
+ * @param  {any}    value     The new value for the setting
+ * @return {Object}           Action object
+ */
+export function setAdvancedSetting( section, setting, value ) {
+	return {
+		type: SET_EXPORTER_ADVANCED_SETTING,
+		section,
+		setting,
+		value
+	};
+}
+
+/**
+ * Ensure that we are fetching or have fetched the available export
+ * settings for the current site.
+ *
+ * @param  {int}      siteId  The ID of the site for which to retrieve export settings
  * @return {Function}         Action thunk
  */
-export function startExport() {
+export function ensureHasSettings() {
+	return ( dispatch, getState ) => {
+		const state = getState();
+		const selectedSite = getSelectedSite( state );
+		const currentDataSiteId = getDataState( state ).forSiteId;
+
+		if ( selectedSite ) {
+			if ( selectedSite.ID !== currentDataSiteId ) {
+				dispatch( fetchExportSettings( selectedSite.ID ) );
+			}
+		}
+	}
+}
+
+/**
+ * Request the available settings for customizing an export.
+ *
+ * @param  {int}      siteId  The ID of the site for which to retrieve export settings
+ * @return {Function}         Action thunk
+ */
+export function fetchExportSettings( siteId ) {
 	return ( dispatch ) => {
 		dispatch( {
-			type: REQUEST_START_EXPORT
+			type: FETCH_EXPORTER_ADVANCED_SETTINGS,
+			siteId: siteId
+		} );
+
+		wpcomUndocumented.getExportSettings( siteId, ( error, data ) => {
+			dispatch( receiveExportSettings( siteId, data ) );
+		} );
+	}
+}
+
+/**
+ * Called when the available export settings are returned from the server.
+ *
+ * @param  {int}      siteId  The ID of the site for which the export settings belong
+ * @param  {Object}   data    The data returned from the server
+ * @return {Function}         Action thunk
+ */
+export function receiveExportSettings( siteId, data ) {
+	return {
+		type: RECEIVE_EXPORTER_ADVANCED_SETTINGS,
+		siteId: siteId,
+		data: data
+	};
+}
+
+/**
+ * Sends a request to the server to start an export.
+ *
+ * @param {string}      postType        The type of post to export, or undefined
+ * @return {Function}                   Action thunk
+ */
+export function startExport( postType ) {
+	// Setting startExport on an event handler passes the event object to
+	// the action creator's first parameter, causing confusing errors. This
+	// assertion will help avoid making this mistake.
+	if ( postType && typeof postType !== 'string' ) {
+		throw new Error(
+			'Expected postType parameter of startExport action creator to ' +
+			'be a string, got ' + ( typeof postType ) + '. Did you pass the action ' +
+			'creator directly to an event handler? Wrap it to make first param blank: ' +
+			'onClick={ ( event ) => startExport() }'
+		);
+	}
```

Comment:
I wrote about this above, but I'm not sure how valuable this is. I _like_ that it's well documented and clear about _why_ it's here, but the other side of the pipe is still opaque. Earlier in this very review I was confused because I didn't realize that calling `startExport` could lead to disaster and there was nothing up there warning me.

The first question I have about this is whether or not it's really a problem. Maybe we saw this while developing the feature and thought that we should make sure we don't do it again, but _will_ we likely do it again and not notice that the whole thing is broken? Actually, if you have already read my earlier comment, I think there's a good reason to write another (possibly a helper) function so that this is never an issue.

Second :question: is about the mixing of error-checking code and logic. There are ten lines of error-checking code for errors that should have all been caught at compile-time (meaning there's no rational reason why these should show up in production if they aren't there in development). The value of the benefits over the cost of maintenance here seems dubious.

Props to the great error message, though!